### PR TITLE
Feat/add flow log level resource

### PR DESCRIPTION
**Pull request by ben-howarth-genesys**
adds support for setting flow log level resource

### CODE REVIEW

File: docs/resources/flow_loglevel.md
Comment (charliecon) [inline docs/resources/flow_loglevel.md:40]: Why is this field listed as a read only attribute? Is this something the user should be able to configure using this resource?

File: examples/resources/genesyscloud_flow_logLevel/apis.md
Comment (dginty4) [reply examples/resources/genesyscloud_flow_logLevel/apis.md:6]: These endpoints are preview apis, you could link to them on the preview api's page on api explorer, https://developer.genesys.cloud/platform/preview-apis

File: genesyscloud/flow_logLevel/resource_genesyscloud_flow_logLevel.go
Comment (dginty4) [inline genesyscloud/flow_logLevel/resource_genesyscloud_flow_logLevel.go:52]: You can probably remove this log, they tend to be unnessary

File: genesyscloud/flow_logLevel/resource_genesyscloud_flow_logLevel.go
Comment (dginty4) [inline genesyscloud/flow_logLevel/resource_genesyscloud_flow_logLevel.go:103]: Instead of getting the value returned from `cc.CheckState()` and printing it, you can do `return cc.CheckState()`

File: genesyscloud/flow_logLevel/resource_genesyscloud_flow_logLevel.go
Comment (dginty4) [inline genesyscloud/flow_logLevel/resource_genesyscloud_flow_logLevel.go:146]: Just a wrong comment here

File: genesyscloud/flow_logLevel/resource_genesyscloud_flow_logLevel_test.go
Comment (dginty4) [inline genesyscloud/flow_logLevel/resource_genesyscloud_flow_logLevel_test.go:116]: Wrong comment here, there's also another a few lines down

File: genesyscloud/flow_logLevel/resource_genesyscloud_flow_logLevel_schema.go
Comment (ben-howarth-genesys) [reply genesyscloud/flow_logLevel/resource_genesyscloud_flow_logLevel_schema.go:110]: No, I'll remove the whole thing

File: docs/resources/flow_loglevel.md
Comment (ben-howarth-genesys) [reply docs/resources/flow_loglevel.md:40]: these are returned by the api and are configured based on the log level that the user sets, I included them as they're illustrative of the log level after an operation but maybe it's something we don't need?

File: genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel_utils.go
Comment (dginty4) [inline genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel_utils.go:10]: This function can be in the test file. We only need to have these functions outside of the test file when they are public since functions can't be public in a test file. In this case though since the function is private it can stay in the test file

File: genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel_schema.go
Comment (charliecon) [inline genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel_schema.go:97]: I would add `MaxItems: 1`  since the user shouldn't define more than one `flow_characteristics` block

File: genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel_utils.go
Comment (charliecon) [inline genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel_utils.go:49]: A panic could be thrown here. Since all the fields in `flow_characteristics` are optional, the user could define the block like this `flow_characteristics{}` , then the list would have one item and it will be nil. 

As a work around, you can just define the map variable like this

```golang
characteristics, ok := d.Get("flow_characteristics").([]interface{})[0].(map[string]interface{})
if !ok {
  return nil
}
```

File: docs/resources/flow_loglevel.md
Comment (charliecon) [reply docs/resources/flow_loglevel.md:39]: No, it looks like we need them. If you can configure them on a POST request, then they belong here as well. I'm just confused as to why they're listed as read only here. Did you use the `go generate` command for these docs? Nothing should be written manually inside the docs folder because this command will always overwrite it

File: genesyscloud/flow_loglevel/genesyscloud_flow_loglevel_proxy.go
Comment (dginty4) [inline genesyscloud/flow_loglevel/genesyscloud_flow_loglevel_proxy.go:1]: There's a few comments in this file that just say external contacts. it might be easier to do a find on the package just to find any incorrect comments and stuff

File: examples/resources/genesyscloud_flow_loglevel/resource.tf
Comment (dginty4) [inline examples/resources/genesyscloud_flow_loglevel/resource.tf:1]: I would add every attribute the user can set to this example. It helps to show the full structure of the resource

File: genesyscloud/flow_loglevel/genesyscloud_flow_loglevel_proxy.go
Comment (dginty4) [inline genesyscloud/flow_loglevel/genesyscloud_flow_loglevel_proxy.go:131]: We are using a new syntax when looping through an array and adding it to another array. You can do this instead
`totalFlowLogLevels = append(totalFlowLogLevels, *flowSettingsResponse.Entities...)`

File: genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel.go
Comment (dginty4) [inline genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel.go:84]: You don't need this check since flow_id is required in the schema this will always exists

File: genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel.go
Comment (dginty4) [inline genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel.go:91]: You don't need to log the error since it's being returned

File: genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel.go
Comment (dginty4) [inline genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel.go:116]: Why are you getting flow log level here?

File: genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel_test.go
Comment (dginty4) [inline genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel_test.go:61]: I think you might need to update these comments on the test steps?

File: genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel_utils.go
Comment (dginty4) [inline genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel_utils.go:32]: I think you can change the logic of this function just to make it shorter. First you can create the `platformclientv2.Flowloglevel` struct with the `Level`. After that do your check to see if you need to add the `Characteristics`, you don't need an else that way and can just have one return statement. P.S. There's a function built into our go sdk that can return a pointer to a string, you can use it for required attributes, e.g. 
```
logLevel := platformclientv2.Flowloglevel{
	Level:   platformclientv2.String(d.Get("flow_log_level").(string)),
}
```

File: genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel_utils.go
Comment (dginty4) [inline genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel_utils.go:25]: This function isn't really doing a lot, It would be easier if you create the `Flowloglevelrequest` directly in the create and update.

File: genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel.go
Comment (ben-howarth-genesys) [reply genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel.go:116]: it's an update method so verifying that the flow exists before taking that action was the idea. I'll remove it though, seems redundant looking at it now

File: examples/resources/genesyscloud_flow_loglevel/resource.tf
Comment (ben-howarth-genesys) [reply examples/resources/genesyscloud_flow_loglevel/resource.tf:1]: every possible flow log level or all the sub fields? The flowLogLevel characteristics can't be set individually, they're just set and returned based on the flow_log_level value a user provides

File: docs/resources/flow_loglevel.md
Comment (ben-howarth-genesys) [reply docs/resources/flow_loglevel.md:39]: you can include them on a post request but they do not have an impact. The characteristics are set by the `flow_log_level` field even though they are accepted by the post request and returned by it. I haven't edited anything here manually but I'll run `go generate` again

File: genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel.go
Comment (dginty4) [reply genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel.go]: I can understand why you thought this, there are a lot of update functions that do a get before the update. the reason they do this is if the genesyscloud object has a version property so we need to do a get to the the current version before updating. In this case though there's no version so you can leave it out

**Comment by dginty4** [comment]
@ben-howarth-genesys Would you be able to clarify something for me please. 
Screenshot 2024-04-11 at 16 29 02 (2)
What properties on the `FlowLogLevel` object should the user set? My understanding from some of the conversations here is that they should only set `level` and the others are read-only? Am I correct?

**Comment by dginty4** [comment]
@ben-howarth-genesys In that case I think you can remove the characteristics attribute from the resource. I understand why you added it but terraform is a tool used to manage infrastructure thus terraform resources only need to include attributes that impact behaviour. For this reason, read-only attributes should not be included unless they are referenced by another resource, but this is rare.  That is my understanding of it anyway, @charliecon  @carnellj-genesys @HemanthDogiparthi12 would you agree or no?

File: genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel_utils.go
Comment (dginty4) [inline genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel_utils.go:1]: This has become very small, you can probably add `getFlowLogLevelFromResourceData` to the resource file resource_genesyscloud_flow_loglevel.go and remove this

File: genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel_test.go
Comment (dginty4) [inline genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel_test.go:53]: This test step will actually update the resource. Resources are not deleted between test steps so for each test step after the first one, the update function is called. You don't need to change the test or anything I would just update the comment to say update, same for the one below. I know this is just a comment but I don't want someone to look at this is the future and be confused.

File: genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel_schema.go
Comment (dginty4) [inline genesyscloud/flow_loglevel/resource_genesyscloud_flow_loglevel_schema.go:41]: You'll need to add `ForceNew: true` to this attribute so that the resource will be recreated if the flow_id changes

File: genesyscloud/flow_loglevel/genesyscloud_flow_loglevel_proxy.go
Comment (dginty4) [inline genesyscloud/flow_loglevel/genesyscloud_flow_loglevel_proxy.go:130]: You're just missing one check here, you need to check if `flowSettingsResponse` is nil or if the length of the `Entities` is 0, there should be example in other resource. You'll also need this check in the loop below

**Comment by dginty4** [comment]
@ben-howarth-genesys I just left a few more comments, after that it should be good. Just one more thing though. There's a test file for the exporter `tfexporter/tf_exporter_resource_test.go`, you will need to add your resource to that file. You'll see two list in the file, add the resource to one list and the exporter to the second

File: genesyscloud/flow_loglevel/genesyscloud_flow_loglevel_proxy.go
Comment (charliecon) [inline genesyscloud/flow_loglevel/genesyscloud_flow_loglevel_proxy.go:133]: One thing before we merge - I don't believe you should return an error here. No entities being returned isn't an error in the API or the terraform provider. You can just `return &totalFlowLogLevels, apiResponse, nil`

File: genesyscloud/flow_loglevel/genesyscloud_flow_loglevel_proxy.go
Comment (charliecon) [inline genesyscloud/flow_loglevel/genesyscloud_flow_loglevel_proxy.go:143]: Same here

**Review by charliecon — APPROVED** [review]

**Review by dginty4 — APPROVED** [review]
There's just a few merge conflicts to resolve and you should be good. I have put `do not merge` on the PR. We just need to enable the feature toggle for these endpoints in our test org before we can merge this.

**Pull request merged**
